Some of mlpack tests are failing on aarch64 machines

Hello everyone,

Following the work on this pull request #2531, we have noticed recently that there are several mlpack tests that are failing when executed on aarch64 architecture (arm architecture 64 bits). These tests are related to:

* FastMKSTest: Failing
* AKFNTest: Failing
* AKNNTest: Failing
* KFNTest: Failing
* KNNTest: Failing
* KMeansTest: Failing (DTNNCoverTreeTest) precisely
* KDETest: Failing (StandardCoverTreeGaussianTest)
* RangeSearchTest: Failing (RSCoverTreeTest)
* EmstTest: Failing (EmstCoverTreeTest)
* KMeansMainTest: Failing (AlgorithmSimilarTest)
* KDEMainTest: Failing (KDEMonoResultMain)

Note that, not all the tests inside these tags are failing, but some of them (Those related to dual cover tree). After running these failing tests, you should expect to enter an infinite loop that never ends. Finally, you will have to use Ctrl-C to stop it.

Usually, we cross-compile binaries from x86_64 bit machines to aarch64 machines, however lately I have found that 
the same issue is produced even though if I compiled these binaries locally on aarch64 (depending on which test you are executing)

I have tested several tests, the following tests (I thought they were relevant) are passing without any issue:
* TreeTest: Passing
* RectangleTreeTraitsTest: Passing
* KernelTest: Passing
* CosineTreeTest: Passing
* KRANNTest: Passing
* HoeffdingTreeTest: Passing
* SerializationTest: Passing
* TreeTraitsTest: Passing
* UBTreeTest: Passing
* VanatagePointTreeTest: Passing

If you have any test to execute that you think might be relevant to this issue, please let me know, and I will add the results back here. To reproduce the issue you need to have access to aarch64 machine through qemu emulator or physical hardware, if you have an idea and you would like to try, but you do not have access to the machine, please leave in a comment and a piece of code to add, modify or to examine and I will be very happy to the tests.

#### Here are some details about the test environment
 * Raspberry Pi 3 with ubuntu 20. (Zin64 on Jenkins)
 * Viking machine on Jenkins
 * version of mlpack: The master version
 * operating system: arch Linux, ubuntu
 * cross-compiler: the mainstream gcc 10 cases more of this issue, I am using this version of buildroot [https://toolchains.bootlin.com/](cross-compilers)
 * compiler: the mainstream version of gcc that you can find in any aarch64

#### Interesting info:
Try to use the buildroot cross-compiler set, since it is the only one that produced binaries that were having the same failure as the local compiler for an aarch64. However, if you use the main gcc cross-compiler. I think it should be fine.  


#### Steps to reproduce
Cross-compile mlpack tests for aarch64 or locally build these tests from aarch64 machine.
`./bin/mlpack_test  [FastMKSTest]`


PS: I will update the issue regularly with any tests that are passing that are using the cover tree

I tried following the directions; thanks for providing those.  I had to do a little bit of adaptation (mostly changing `TOOLCHAIN_PREFIX`), but nothing too serious.  I also had to download the dependencies and install them into the right place, but it seems like maybe something has been linked wrong?  I get this when I build:

```
...
/home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/../../../../aarch64-buildroot-linux-gnu/bin/ld: /home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/libgcc_eh.a(unwind-dw2-fde-dip.o): in function `__gthread_mutex_lock':
/opt/output/build/host-gcc-final-9.3.0/build/aarch64-buildroot-linux-gnu/libgcc/./gthr-default.h:749: undefined reference to `pthread_mutex_lock'
/home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/../../../../aarch64-buildroot-linux-gnu/bin/ld: /home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/libgcc_eh.a(unwind-dw2-fde-dip.o): in function `__gthread_mutex_unlock':
/opt/output/build/host-gcc-final-9.3.0/build/aarch64-buildroot-linux-gnu/libgcc/./gthr-default.h:779: undefined reference to `pthread_mutex_unlock'
/home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/../../../../aarch64-buildroot-linux-gnu/bin/ld: /home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/libgcc_eh.a(unwind-dw2-fde-dip.o): in function `__gthread_mutex_lock':
/opt/output/build/host-gcc-final-9.3.0/build/aarch64-buildroot-linux-gnu/libgcc/./gthr-default.h:749: undefined reference to `pthread_mutex_lock'
/home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/../../../../aarch64-buildroot-linux-gnu/bin/ld: /home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/libgcc_eh.a(unwind-dw2-fde-dip.o): in function `__gthread_mutex_unlock':
/opt/output/build/host-gcc-final-9.3.0/build/aarch64-buildroot-linux-gnu/libgcc/./gthr-default.h:779: undefined reference to `pthread_mutex_unlock'
/home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/../../../../aarch64-buildroot-linux-gnu/bin/ld: /home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/libgcc_eh.a(unwind-dw2-fde-dip.o): in function `__gthread_mutex_lock':
/opt/output/build/host-gcc-final-9.3.0/build/aarch64-buildroot-linux-gnu/libgcc/./gthr-default.h:749: undefined reference to `pthread_mutex_lock'
/home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/../../../../aarch64-buildroot-linux-gnu/bin/ld: /home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/libgcc_eh.a(unwind-dw2-fde-dip.o): in function `__gthread_mutex_unlock':
/opt/output/build/host-gcc-final-9.3.0/build/aarch64-buildroot-linux-gnu/libgcc/./gthr-default.h:779: undefined reference to `pthread_mutex_unlock'
/home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/../../../../aarch64-buildroot-linux-gnu/bin/ld: /home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/libgcc_eh.a(unwind-dw2-fde-dip.o): in function `__gthread_mutex_lock':
/opt/output/build/host-gcc-final-9.3.0/build/aarch64-buildroot-linux-gnu/libgcc/./gthr-default.h:749: undefined reference to `pthread_mutex_lock'
/home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/../../../../aarch64-buildroot-linux-gnu/bin/ld: /home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/libgcc_eh.a(unwind-dw2-fde-dip.o): in function `__gthread_mutex_unlock':
/opt/output/build/host-gcc-final-9.3.0/build/aarch64-buildroot-linux-gnu/libgcc/./gthr-default.h:779: undefined reference to `pthread_mutex_unlock'
collect2: error: ld returned 1 exit status
make[2]: *** [src/mlpack/methods/cf/CMakeFiles/mlpack_cf.dir/build.make:106: bin/mlpack_cf] Error 1
make[1]: *** [CMakeFiles/Makefile2:2810: src/mlpack/methods/cf/CMakeFiles/mlpack_cf.dir/all] Error 2
make: *** [Makefile:160: all] Error 2
```

--- a/src/mlpack/core/tree/cover_tree/dual_tree_traverser.hpp
+++ b/src/mlpack/core/tree/cover_tree/dual_tree_traverser.hpp
@@ -100,7 +100,8 @@ class CoverTree<MetricType, StatisticType, MatType, RootPointPolicy>::
                          std::vector<DualCoverTreeMapEntry>,
                          std::greater<int>>& childMap);
 
-  void ReferenceRecursion(CoverTree& queryNode,
+  void ReferenceRecursion(
+      CoverTree& queryNode,
     std::map<int, std::vector<DualCoverTreeMapEntry>,
         std::greater<int>>& referenceMap);
 };

--- a/src/mlpack/core/tree/cover_tree/dual_tree_traverser.hpp
+++ b/src/mlpack/core/tree/cover_tree/dual_tree_traverser.hpp
@@ -86,7 +86,8 @@ class CoverTree<MetricType, StatisticType, MatType, RootPointPolicy>::
   /**
    * Helper function for traversal of the two trees.
    */
-  void Traverse(CoverTree& queryNode,
+  void Traverse(
+      CoverTree& queryNode,
       std::map<int, std::vector<DualCoverTreeMapEntry>,
           std::greater<int>>& referenceMap);
 

--- a/src/mlpack/core/tree/cover_tree/dual_tree_traverser.hpp
+++ b/src/mlpack/core/tree/cover_tree/dual_tree_traverser.hpp
@@ -92,7 +92,8 @@ class CoverTree<MetricType, StatisticType, MatType, RootPointPolicy>::
                          std::greater<int>>& referenceMap);
 
   //! Prepare map for recursion.
-  void PruneMap(CoverTree& queryNode,
+  void PruneMap(
+      CoverTree& queryNode,
       std::map<int, std::vector<DualCoverTreeMapEntry>,
           std::greater<int>>& referenceMap,
       std::map<int, std::vector<DualCoverTreeMapEntry>,

--- a/src/mlpack/core/tree/cover_tree/dual_tree_traverser.hpp
+++ b/src/mlpack/core/tree/cover_tree/dual_tree_traverser.hpp
@@ -101,9 +101,8 @@ class CoverTree<MetricType, StatisticType, MatType, RootPointPolicy>::
                          std::greater<int>>& childMap);
 
   void ReferenceRecursion(CoverTree& queryNode,
-                          std::map<int,
-                                   std::vector<DualCoverTreeMapEntry>,
-                                   std::greater<int>>& referenceMap);
+    std::map<int, std::vector<DualCoverTreeMapEntry>,
+        std::greater<int>>& referenceMap);
 };
 
 } // namespace tree

--- a/src/mlpack/core/tree/cover_tree/dual_tree_traverser.hpp
+++ b/src/mlpack/core/tree/cover_tree/dual_tree_traverser.hpp
@@ -93,12 +93,10 @@ class CoverTree<MetricType, StatisticType, MatType, RootPointPolicy>::
 
   //! Prepare map for recursion.
   void PruneMap(CoverTree& queryNode,
-                std::map<int,
-                         std::vector<DualCoverTreeMapEntry>,
-                         std::greater<int>>& referenceMap,
-                std::map<int,
-                         std::vector<DualCoverTreeMapEntry>,
-                         std::greater<int>>& childMap);
+      std::map<int, std::vector<DualCoverTreeMapEntry>,
+          std::greater<int>>& referenceMap,
+      std::map<int, std::vector<DualCoverTreeMapEntry>,
+          std::greater<int>>& childMap);
 
   void ReferenceRecursion(
       CoverTree& queryNode,

--- a/src/mlpack/core/tree/cover_tree/dual_tree_traverser.hpp
+++ b/src/mlpack/core/tree/cover_tree/dual_tree_traverser.hpp
@@ -87,19 +87,23 @@ class CoverTree<MetricType, StatisticType, MatType, RootPointPolicy>::
    * Helper function for traversal of the two trees.
    */
   void Traverse(CoverTree& queryNode,
-                std::map<int, std::vector<DualCoverTreeMapEntry> >&
-                    referenceMap);
+                std::map<int,
+                         std::vector<DualCoverTreeMapEntry>,
+                         std::greater<int>>& referenceMap);
 
   //! Prepare map for recursion.
   void PruneMap(CoverTree& queryNode,
-                std::map<int, std::vector<DualCoverTreeMapEntry> >&
-                    referenceMap,
-                std::map<int, std::vector<DualCoverTreeMapEntry> >&
-                    childMap);
+                std::map<int,
+                         std::vector<DualCoverTreeMapEntry>,
+                         std::greater<int>>& referenceMap,
+                std::map<int,
+                         std::vector<DualCoverTreeMapEntry>,
+                         std::greater<int>>& childMap);
 
   void ReferenceRecursion(CoverTree& queryNode,
-                          std::map<int, std::vector<DualCoverTreeMapEntry> >&
-                              referenceMap);
+                          std::map<int,
+                                   std::vector<DualCoverTreeMapEntry>,
+                                   std::greater<int>>& referenceMap);
 };
 
 } // namespace tree

--- a/src/mlpack/core/tree/cover_tree/dual_tree_traverser.hpp
+++ b/src/mlpack/core/tree/cover_tree/dual_tree_traverser.hpp
@@ -87,9 +87,8 @@ class CoverTree<MetricType, StatisticType, MatType, RootPointPolicy>::
    * Helper function for traversal of the two trees.
    */
   void Traverse(CoverTree& queryNode,
-                std::map<int,
-                         std::vector<DualCoverTreeMapEntry>,
-                         std::greater<int>>& referenceMap);
+      std::map<int, std::vector<DualCoverTreeMapEntry>,
+          std::greater<int>>& referenceMap);
 
   //! Prepare map for recursion.
   void PruneMap(

--- a/src/mlpack/core/tree/cover_tree/dual_tree_traverser_impl.hpp
+++ b/src/mlpack/core/tree/cover_tree/dual_tree_traverser_impl.hpp
@@ -43,7 +43,7 @@ DualTreeTraverser<RuleType>::Traverse(CoverTree& queryNode,
                                       CoverTree& referenceNode)
 {
   // Start by creating a map and adding the reference root node to it.
-  std::map<int, std::vector<DualCoverTreeMapEntry> > refMap;
+  std::map<int, std::vector<DualCoverTreeMapEntry>, std::greater<int>> refMap;
 
   DualCoverTreeMapEntry rootRefEntry;
 
@@ -70,7 +70,8 @@ template<typename RuleType>
 void CoverTree<MetricType, StatisticType, MatType, RootPointPolicy>::
 DualTreeTraverser<RuleType>::Traverse(
     CoverTree& queryNode,
-    std::map<int, std::vector<DualCoverTreeMapEntry> >& referenceMap)
+    std::map<int, std::vector<DualCoverTreeMapEntry>, std::greater<int>>&
+        referenceMap)
 {
   if (referenceMap.size() == 0)
     return; // Nothing to do!
@@ -85,7 +86,7 @@ DualTreeTraverser<RuleType>::Traverse(
   // Now, reduce the scale of the query node by recursing.  But we can't recurse
   // if the query node is a leaf node.
   if ((queryNode.Scale() != INT_MIN) &&
-      (queryNode.Scale() >= (*referenceMap.rbegin()).first))
+      (queryNode.Scale() >= (*referenceMap.begin()).first))
   {
     // Recurse into the non-self-children first.  The recursion order cannot
     // affect the runtime of the algorithm, because each query child recursion's
@@ -95,11 +96,15 @@ DualTreeTraverser<RuleType>::Traverse(
     for (size_t i = 1; i < queryNode.NumChildren(); ++i)
     {
       // We need a copy of the map for this child.
-      std::map<int, std::vector<DualCoverTreeMapEntry> > childMap;
+      std::map<int, std::vector<DualCoverTreeMapEntry>, std::greater<int>>
+          childMap;
+
       PruneMap(queryNode.Child(i), referenceMap, childMap);
       Traverse(queryNode.Child(i), childMap);
     }
-    std::map<int, std::vector<DualCoverTreeMapEntry> > selfChildMap;
+    std::map<int, std::vector<DualCoverTreeMapEntry>, std::greater<int>>
+        selfChildMap;
+
     PruneMap(queryNode.Child(0), referenceMap, selfChildMap);
     Traverse(queryNode.Child(0), selfChildMap);
   }
@@ -111,8 +116,7 @@ DualTreeTraverser<RuleType>::Traverse(
   // evaluations to do.
   Log::Assert((*referenceMap.begin()).first == INT_MIN);
   Log::Assert(queryNode.Scale() == INT_MIN);
-  std::vector<DualCoverTreeMapEntry>& pointVector =
-      (*referenceMap.begin()).second;
+  std::vector<DualCoverTreeMapEntry>& pointVector = referenceMap[INT_MIN];
 
   for (size_t i = 0; i < pointVector.size(); ++i)
   {
@@ -156,25 +160,25 @@ template<typename RuleType>
 void CoverTree<MetricType, StatisticType, MatType, RootPointPolicy>::
 DualTreeTraverser<RuleType>::PruneMap(
     CoverTree& queryNode,
-    std::map<int, std::vector<DualCoverTreeMapEntry> >& referenceMap,
-    std::map<int, std::vector<DualCoverTreeMapEntry> >& childMap)
+    std::map<int, std::vector<DualCoverTreeMapEntry>, std::greater<int>>&
+        referenceMap,
+    std::map<int, std::vector<DualCoverTreeMapEntry>, std::greater<int>>&
+        childMap)
 {
   if (referenceMap.empty())
     return; // Nothing to do.
 
   // Copy the zero set first.
-  if ((*referenceMap.begin()).first == INT_MIN)
+  if (referenceMap.count(INT_MIN) == 1)
   {
     // Get a reference to the vector representing the entries at this scale.
-    std::vector<DualCoverTreeMapEntry>& scaleVector =
-        (*referenceMap.begin()).second;
+    std::vector<DualCoverTreeMapEntry>& scaleVector = referenceMap[INT_MIN];
 
     // Before traversing all the points in this scale, sort by score.
     std::sort(scaleVector.begin(), scaleVector.end());
 
-    const int thisScale = (*referenceMap.begin()).first;
-    childMap[thisScale].reserve(scaleVector.size());
-    std::vector<DualCoverTreeMapEntry>& newScaleVector = childMap[thisScale];
+    childMap[INT_MIN].reserve(scaleVector.size());
+    std::vector<DualCoverTreeMapEntry>& newScaleVector = childMap[INT_MIN];
 
     // Loop over each entry in the vector.
     for (size_t j = 0; j < scaleVector.size(); ++j)
@@ -208,13 +212,13 @@ DualTreeTraverser<RuleType>::PruneMap(
 
     // If we didn't add anything, then strike this vector from the map.
     if (newScaleVector.size() == 0)
-      childMap.erase((*referenceMap.begin()).first);
+      childMap.erase(INT_MIN);
   }
 
-  typename std::map<int, std::vector<DualCoverTreeMapEntry> >::reverse_iterator
-      it = referenceMap.rbegin();
+  typename std::map<int, std::vector<DualCoverTreeMapEntry>,
+      std::greater<int>>::iterator it = referenceMap.begin();
 
-  while ((it != referenceMap.rend()))
+  while ((it != referenceMap.end()))
   {
     const int thisScale = (*it).first;
     if (thisScale == INT_MIN) // We already did it.
@@ -277,28 +281,26 @@ template<typename RuleType>
 void CoverTree<MetricType, StatisticType, MatType, RootPointPolicy>::
 DualTreeTraverser<RuleType>::ReferenceRecursion(
     CoverTree& queryNode,
-    std::map<int, std::vector<DualCoverTreeMapEntry> >& referenceMap)
+    std::map<int, std::vector<DualCoverTreeMapEntry>, std::greater<int>>&
+        referenceMap)
 {
   // First, reduce the maximum scale in the reference map down to the scale of
   // the query node.
   while (!referenceMap.empty())
   {
+    const int maxScale = ((*referenceMap.begin()).first);
     // Hacky bullshit to imitate jl cover tree.
-    if (queryNode.Parent() == NULL && (*referenceMap.rbegin()).first <
-        queryNode.Scale())
+    if (queryNode.Parent() == NULL && maxScale < queryNode.Scale())
       break;
-    if (queryNode.Parent() != NULL && (*referenceMap.rbegin()).first <=
-        queryNode.Scale())
+    if (queryNode.Parent() != NULL && maxScale <= queryNode.Scale())
       break;
     // If the query node's scale is INT_MIN and the reference map's maximum
     // scale is INT_MIN, don't try to recurse...
-    if ((queryNode.Scale() == INT_MIN) &&
-       ((*referenceMap.rbegin()).first == INT_MIN))
+    if (queryNode.Scale() == INT_MIN && maxScale == INT_MIN)
       break;
 
     // Get a reference to the current largest scale.
-    std::vector<DualCoverTreeMapEntry>& scaleVector =
-        (*referenceMap.rbegin()).second;
+    std::vector<DualCoverTreeMapEntry>& scaleVector = referenceMap[maxScale];
 
     // Before traversing all the points in this scale, sort by score.
     std::sort(scaleVector.begin(), scaleVector.end());
@@ -308,7 +310,6 @@ DualTreeTraverser<RuleType>::ReferenceRecursion(
     {
       // Get a reference to the current element.
       const DualCoverTreeMapEntry& frame = scaleVector.at(i);
-
       CoverTree* refNode = frame.referenceNode;
 
       // Create the score for the children.
@@ -344,13 +345,12 @@ DualTreeTraverser<RuleType>::ReferenceRecursion(
         newFrame.score = childScore; // Use the score of the parent.
         newFrame.baseCase = baseCase;
         newFrame.traversalInfo = rule.TraversalInfo();
-
         referenceMap[newFrame.referenceNode->Scale()].push_back(newFrame);
       }
     }
 
     // Now clear the memory for this scale; it isn't needed anymore.
-    referenceMap.erase((*referenceMap.rbegin()).first);
+    referenceMap.erase(maxScale);
   }
 }
 

--- a/src/mlpack/core/tree/cover_tree/single_tree_traverser_impl.hpp
+++ b/src/mlpack/core/tree/cover_tree/single_tree_traverser_impl.hpp
@@ -80,9 +80,9 @@ SingleTreeTraverser<RuleType>::Traverse(
   // and then the vector is all the nodes in that scale which need to be
   // investigated.  Because no point in a scale can add a point in its own
   // scale, we know that the vector for each scale is final when we get to it.
-  // In addition, map is organized in such a way that rbegin() will return the
-  // largest scale.
-  std::map<int, std::vector<MapEntryType> > mapQueue;
+  // In addition, the map is organized in such a way that begin() will return
+  // the largest scale.
+  std::map<int, std::vector<MapEntryType>, std::greater<int>> mapQueue;
 
   // Create the score for the children.
   double rootChildScore = rule.Score(queryIndex, referenceNode);
@@ -123,14 +123,13 @@ SingleTreeTraverser<RuleType>::Traverse(
   // Now begin the iteration through the map, but only if it has anything in it.
   if (mapQueue.empty())
     return;
-  typename std::map<int, std::vector<MapEntryType> >::reverse_iterator rit =
-      mapQueue.rbegin();
+  int maxScale = mapQueue.cbegin()->first;
 
   // We will treat the leaves differently (below).
-  while ((*rit).first != INT_MIN)
+  while (maxScale != INT_MIN)
   {
     // Get a reference to the current scale.
-    std::vector<MapEntryType>& scaleVector = (*rit).second;
+    std::vector<MapEntryType>& scaleVector = mapQueue[maxScale];
 
     // Before traversing all the points in this scale, sort by score.
     std::sort(scaleVector.begin(), scaleVector.end());
@@ -170,7 +169,9 @@ SingleTreeTraverser<RuleType>::Traverse(
       // trees using TreeTraits::FirstPointIsCentroid; this is an optimization
       // that (theoretically) the compiler should get right.
       if (point != parent)
+      {
         baseCase = rule.BaseCase(queryIndex, point);
+      }
 
       // Don't add the self-leaf.
       size_t j = 0;
@@ -193,7 +194,8 @@ SingleTreeTraverser<RuleType>::Traverse(
     }
 
     // Now clear the memory for this scale; it isn't needed anymore.
-    mapQueue.erase((*rit).first);
+    mapQueue.erase(maxScale);
+    maxScale = mapQueue.begin()->first;
   }
 
   // Now deal with the leaves.

--- a/src/mlpack/tests/akfn_test.cpp
+++ b/src/mlpack/tests/akfn_test.cpp
@@ -241,4 +241,3 @@ TEST_CASE("AKFNDualBallTreeTest", "[AKFNTest]")
   for (size_t i = 0; i < neighborsBallTree.n_elem; ++i)
     REQUIRE_RELATIVE_ERR(distancesBallTree(i), distancesExact(i), 0.05);
 }
-

--- a/src/mlpack/tests/krann_search_test.cpp
+++ b/src/mlpack/tests/krann_search_test.cpp
@@ -368,7 +368,8 @@ TEST_CASE("DualCoverTreeTest", "[KRANNTest]")
   RACoverTreeSearch tsdRann(&refTree, false, 1.0, 0.95, false, false, 5);
 
   arma::Mat<size_t> qrRanks;
-  if (!data::Load("rann_test_qr_ranks.csv", qrRanks, false, false)) // No transpose.
+  // No transpose.
+  if (!data::Load("rann_test_qr_ranks.csv", qrRanks, false, false))
     FAIL("Cannot load dataset rann_test_qr_ranks.csv");
 
   size_t numRounds = 100;

